src/lib.rs

```diff
@@ -0,0 +1,222 @@
+use pyo3::ffi::{
+    PyContextVar_Get, PyEval_SetProfile, PyFrameObject, PyUnicode_AsUTF8, Py_tracefunc,
+    CO_ASYNC_GENERATOR, CO_COROUTINE, CO_ITERABLE_COROUTINE,
+};
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use pyo3::{AsPyPointer, PyAny, Python};
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use std::ffi::CStr;
+use std::os::raw::c_int;
+use std::time::Instant;
+
+/// Enum of possible Python's Trace/Profiling events
+#[allow(dead_code)]
+#[repr(i32)]
+enum TraceEvent {
+    Call,
+    Exception,
+    Line,
+    Return,
+    CCall,
+    CException,
+    CReturn,
+    Opcode,
+    __NonExhaustive,
+}
+
+impl TryFrom<c_int> for TraceEvent {
+    type Error = &'static str;
+    /// Cast i32 event (raw from Python) to Rust enum.
+    fn try_from(value: i32) -> Result<Self, Self::Error> {
+        if value > 7 || value < 0 {
+            return Err("Not valid enum value");
+        }
+        Ok(unsafe { std::mem::transmute(value) })
+    }
+}
+
+/// Struct representing frame data
+/// Identifier is frame pointer casted to usize.
+struct FrameData {
+    func_name: String,
+    file_name: String,
+    identifier: usize,
+    yielded_coroutine: bool,
+}
+
+impl TryFrom<*mut PyFrameObject> for FrameData {
+    type Error = &'static str;
+    fn try_from(frame: *mut PyFrameObject) -> Result<Self, Self::Error> {
+        if frame.is_null() {
+            return Err("frame is null");
+        }
+
+        let dframe = unsafe { *frame };
+
+        if dframe.f_code.is_null() {
+            return Err("f_code is null");
+        }
+
+        let code = unsafe { *dframe.f_code };
+        let yielded_coroutine = {
+            (0 < (code.co_flags & CO_COROUTINE)
+                | (code.co_flags & CO_ITERABLE_COROUTINE)
+                | (code.co_flags & CO_ASYNC_GENERATOR))
+                && (!dframe.f_stacktop.is_null())
+        };
+
+        unsafe {
+            let file_name = match code.co_filename.is_null() {
+                true => "null".to_string(),
+                false => CStr::from_ptr(PyUnicode_AsUTF8(code.co_filename))
+                    .to_string_lossy()
+                    .into_owned(),
+            };
```

Comment:
Consider extracting this bit into a function. It's repeated in the following line so it would do good for code reuse as well as better constraining the `unsafe` code